DESIGN.md

```diff
@@ -0,0 +1,27 @@
+## Overview
+
+### Target Audience
+
+- Developer A: high-level framework writer
+  - wants to provide barebones initial configuration
+  - wants to consume a simple event stream
+  - wants to inject events into a managed event loop (easy state machine semantics)
+
+```go
+c := client.New(connectionOptions..) // package scheduler/client
+eventCh, errCh := c.Register(registrationOptions..)
```

Comment:
Yes, the example scheduler code that you posted fits very well with what I
had in mind for "Developer B".

However there's some common helpful things that mesos-go could provide as
well for Developer A -- but I don't want to necessarily force every
framework writer to use them:

- basic implementation of the event loop as part of a simple state machine,
that;
- manages auto-reconnection/re-registration to the latest mesos master
- sends appropriate events back to the scheduler in the event of a
disconnect during a pending call (e.g. taskLaunch)
- additional support for management of reserved resources (haven't played
with these, but from what I've seen there are some useful patterns to adopt
here)

Developer B can handle all of this themselves if they want to. But it would
be nice to eventually implement higher-level support things for Developer A
as well. We can certainly focus on the Developer B stuff first.


On Fri, Nov 13, 2015 at 5:02 AM, Tomás Senart <notifications@github.com>
wrote:

> In DESIGN.md
> <https://github.com/mesos/mesos-go/pull/210#discussion_r44766491>:
>
> > @@ -0,0 +1,27 @@
> > +## Overview
> > +
> > +### Target Audience
> > +
> > +- Developer A: high-level framework writer
> > +  - wants to provide barebones initial configuration
> > +  - wants to consume a simple event stream
> > +  - wants to inject events into a managed event loop (easy state machine semantics)
> > +
> > +```go
> > +c := client.New(connectionOptions..) // package scheduler/client
> > +eventCh, errCh := c.Register(registrationOptions..)
>
> +1
>
> The original idea was to have schedulers just decode events in a for loop.
> I realised I had some lingering code in my computer which I just pushed to
> #211 <https://github.com/mesos/mesos-go/pull/211>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesos/mesos-go/pull/210/files#r44766491>.
>